modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1285,3 +1285,103 @@ def finalize(cls, partitions):
+            The interval is determined by the chunksize.
+
+            Parameters
+            ----------
+            axis_lengths : list
+                The length/width of each partition in the rows/columns.
+            axis : {0, 1}
+                Axis for creating the repartition scheme.
+
+            Returns
+            -------
+            tuple
+                Tuple of (schema, updated axis lengths aсcording to the schema)
+            """
+            idx = pandas.RangeIndex(sum(axis_lengths))
+            # It will be possible to use axis length for `compute_chunksize` instead of dataframe
+            # after merge of https://github.com/modin-project/modin/issues/3768 happens.
+            df = pandas.DataFrame(**{"columns" if axis == 1 else "index": idx})
+            chunksize = compute_chunksize(df, NPartitions.get(), axis=axis)
+
+            schema = []
+            current_axis_length = 0
+
+            for idx, axis_length in enumerate(axis_lengths):
+                if current_axis_length == 0:
+                    # create new interval
+                    schema.append([idx, None])
+                if current_axis_length + axis_length < chunksize:
+                    current_axis_length += axis_length
+                    continue
+                current_axis_length = 0
+                # update end of the interval
+                schema[-1][1] = idx + 1
+
+            if schema[-1][1] is None:
+                schema[-1][1] = len(axis_lengths)
+
+            # update axis lengths aсcording to the schema
+            new_axis_lengths = [sum(axis_lengths[t[0] : t[1]]) for t in schema]
+            return schema, new_axis_lengths
+
+        axis_lengths = widths if axis == 1 else lengths
+        repartition_schema, axis_lengths = compute_schema(axis_lengths, axis)
+
+        new_lengths = axis_lengths if axis == 0 else lengths
+        new_widths = axis_lengths if axis == 1 else widths
+
+        new_partitions = []
+        for start, end in repartition_schema:
+            _parts = partitions[:, start:end] if axis == 1 else partitions[start:end]
+            if end - start == 1:
+                # Fast path; there is only one column/row partition in the merge interval
+                new_partitions.append(_parts)
+                continue
+            new_partitions.append(
+                cls.map_axis_partitions(
+                    axis=axis,
+                    partitions=_parts,
+                    map_func=lambda df: df,
+                    keep_partitioning=False,
+                    # The passed length must be greater than the sum of the merging
+                    # partitions lengths, so as not to lose some of the axis elements.
+                    # It is simply convenient to use the full axis length, since it
+                    # exactly fits the constraint.
+                    lengths=[sum(axis_lengths)],
```

Comment:
Isn't `end - start` enough?

Reply:
These variables represent the indexes of the partitions and do not refer to the index of the dataframe itself, so they can't be used like this.